### PR TITLE
Rscontext vbet buffer (PRISM)

### PR DESCRIPTION
**Pull request by KellyMWhitehead**
This will add support to RS Context for prism data.

1.  Please add the prism geopackage to the National Project under 'prism/Prism.gpkg' (or modify location as needed). This file may need to be updated if new prism attributes are added.

2.  Note the changes in the arguments adding prism.gpkg. These are also reflected in the launch.json file.

### CODE REVIEW

File: lib/commons/rscommons/prism.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+if __name__ == "__main__":
+
+    precip = mean_area_precip(r"D:\NAR_Data\Data\rs_context\17060304\hydrology\WBDHU8.shp", r"D:\NAR\Projects\VBET\Prism_Test02.gpkg")
+
+    calculate_bankfull_width(r"D:\NAR\Projects\VBET\NHDFlowlineCopy.shp", precip)
+
+    polygon = buffer_by_field(r"D:\NAR\Projects\VBET\NHDFlowlineCopy.shp", "BFbuffer")
+
+    print(polygon.area)
```
Comment (MattReimer) [inline lib/commons/rscommons/prism.py:162]: Cab we load this stuff from a .env file? We should avoid any hardcoded paths

File: lib/commons/rscommons/prism.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
+import csv
+import statistics
+from osgeo import ogr
+from shapely.wkt import loads as wkt_load
+from shapely.geometry import mapping, shape, Polygon, MultiPolygon
+from shapely.ops import unary_union
+from rscommons.shapefile import get_transform_from_epsg
+
+def load_mean_annual_data(gpkg, input_csv, attribute_name):
+    '''Loads PRISM attribute csv file into PRISM geopackage as mean annual value'''
```
Comment (MattReimer) [inline lib/commons/rscommons/prism.py:10]: Really small thing but can we get automatic code comments on all the functions?

In VSCode if you type """ and then hit enter it usually fills it in for you.

If that doesn't work open up the workspace setting in VSCode and make sure `docstringFormat` is set to `google`.

```
"autoDocstring.docstringFormat": "google"
"autoDocstring.quoteStyle": "\"\"\""
```

So you should get something like:

```python
def buffer_by_field(flowlines, field, epsg, conversion=1):
    """[summary]

    Args:
        flowlines ([type]): [description]
        field ([type]): [description]
        epsg ([type]): [description]
        conversion (int, optional): [description]. Defaults to 1.

    Returns:
        [type]: [description]
    """

    driver = ogr.GetDriverByName("ESRI Shapefile")
```

File: packages/rscontext/rscontext/rs_context.py [hunk 3]
```diff
@@ -108,6 +110,9 @@ def rs_context(huc, existing_veg, historic_veg, ownership, ecoregions, output_fo
     nhd, db_path, huc_name = clean_nhd_data(huc, nhd_download_folder, nhd_unzip_folder, os.path.join(output_folder, 'hydrology'), cfg.OUTPUT_EPSG, False)
     project.add_metadata({'Watershed': huc_name})
 
+    precip = mean_area_precip(nhd['WBDHU{}'.format(len(huc))], prism)
```
Comment (MattReimer) [inline packages/rscontext/rscontext/rs_context.py:113]: can we set up the geopackage as a project layer the same way the others work?

https://github.com/Riverscapes/riverscapes-tools/blob/master/packages/rscontext/rscontext/rs_context.py#L40-L51

This is going to make it easier to write to the project.rs.xml file.

**Review by MattReimer — CHANGES_REQUESTED** [review]
I wrote some suggestions for some really minor tweaks. Let me know if anything doesn't make sense

File: lib/commons/rscommons/prism.py (hunk 1 above)
Comment (KellyMWhitehead) [reply lib/commons/rscommons/prism.py:162]: I don't actually need this anymore. I will make sure I remove it in the future.

File: lib/commons/rscommons/prism.py (hunk 2 above)
Comment (KellyMWhitehead) [reply lib/commons/rscommons/prism.py:10]: Done for prism.py, will continue to do so in future.

File: packages/rscontext/rscontext/rs_context.py (hunk 3 above)
Comment (KellyMWhitehead) [reply packages/rscontext/rscontext/rs_context.py:113]: Added as a new layer type. I'm not sure what else needs to be done with this for now.

File: packages/rscontext/rscontext/rs_context.py [hunk 4]
```diff
@@ -107,6 +118,21 @@ def rs_context(huc, existing_veg, historic_veg, ownership, ecoregions, output_fo
 
     nhd, db_path, huc_name = clean_nhd_data(huc, nhd_download_folder, nhd_unzip_folder, os.path.join(output_folder, 'hydrology'), cfg.OUTPUT_EPSG, False)
     project.add_metadata({'Watershed': huc_name})
+    boundary = 'WBDHU{}'.format(len(huc))
+
+    # PRISM climate rasters
+    mean_annual_precip = None
+    for variable in ['PPT', 'TMEAN', 'TMIN', 'TMAX', 'TDMEAN', 'VPDMIN', 'VPDMAX']:
+        file_name = 'PRISM_{}_30yr_normal_800mM2_annual_bil'.format(variable.lower())
```
Comment (MattReimer) [inline packages/rscontext/rscontext/rs_context.py]: I'm really not loving this lookup.

File: packages/rscontext/.vscode/launch.json [hunk 5]
```diff
@@ -27,6 +27,7 @@
         "{env:NATIONAL_PROJECT}/landfire/200/us_200bps.tif",
         "{env:NATIONAL_PROJECT}/ownership/surface_management_agency.shp",
         "{env:NATIONAL_PROJECT}/ecoregions/us_eco_l3_state_boundaries.shp",
+        "{env:NATIONAL_PROJECT}/prism/Prism.gpkg",
```
Comment (MattReimer) [inline packages/rscontext/.vscode/launch.json]: This seems wrong. I'll fix it.

**Review by MattReimer — CHANGES_REQUESTED** [review]
Couple of things.

**Review by MattReimer — APPROVED** [review]

**Comment by MattReimer** [comment]
@philipbaileynar @KellyMWhitehead I've merged this using "squash" since the branch got a little messy

```
git fetch -p
git checkout master
git pull
```

Then you can remove your local branch

**Comment by philipbaileynar** [comment]
Thanks @MattReimer 

So now we have PRISM data that is not inside the national project, we have folder path in launch.json that is user specific depending on where you unzip the PRISM data.

Are you OK if I create a new environment variable and add the folder path so that we don't end up wrestling over launch.json in git?

**Pull request merged**